torch/contrib/_tensorboard_vis.py

```diff
@@ -0,0 +1,141 @@
+import time
+from collections import defaultdict
+from functools import partial
+
+
+# Unfortunately it doesn't seem as if there was any way to get TensorBoard to do
+# anything without having TF installed, and so this file has a hard dependency on it
+# as well. It really is a debugging tool, so it doesn't matter.
+try:
+    from tensorflow.core.util import event_pb2
+    from tensorflow.core.framework import graph_pb2
+    from tensorflow.python.summary.writer.writer import FileWriter
+except ImportError:
+    raise ImportError("TensorBoard visualization of GraphExecutors requires having "
+                      "TensorFlow installed")
+
+
+def dump_tensorboard_summary(graph_executor, logdir):
+    with FileWriter(logdir) as w:
+        pb_graph = visualize(graph_executor)
+        evt = event_pb2.Event(wall_time=time.time(), graph_def=pb_graph.SerializeToString())
+        w.add_event(evt)
+
+
+def visualize(graph, name_prefix='', pb_graph=None, executors_it=None):
+    """Visualizes an independent graph, or a graph executor."""
+    value_map = {}
+    pb_graph = pb_graph or graph_pb2.GraphDef()
+
+    if isinstance(graph, (torch._C.GraphExecutor, torch._C.GraphExecutorState)):
+        visualize_graph_executor(graph, name_prefix, pb_graph,
```

Comment:
The interaction between `visualize_graph_executor`'s `inline_graph` field and `visualize` is hard for me to follow.  If there is a way to simplify it that would be nice, but I don't want to block getting this debug code merged just because of this.

Reply:
I agree. The issue is that `visualize_graph_executor` can be called either as a top-level thing, that doesn't have the input/output nodes created just yet, but it can also appear as a nested part of the graph. Both of those cases have to be handled differently if we want to display them as they are right now.